### ISSUE TITLE
`client.Close` sometimes hangs on network error

### ISSUE THREAD

**Issue by tomkcook** [gammazero/nexus#189]
I have a situation where the network upstream of the device running nexus code is unreliable.  When the network connection breaks, it does so without anything telling my nexus-based client that it's broken (ie no TCP FIN or RST packets are sent).  In this case, about one time in two, I see my client hanging in `client.Close()` with this included in the stack (note that line numbers here are for commit 4debb4a but I believe the same defect is present in the current master):

```
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: goroutine 1 [semacquire, 31 minutes]:
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: sync.runtime_Semacquire(0x11a9c560)
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: 	/usr/local/go/src/runtime/sema.go:56 +0x2c
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: sync.(*WaitGroup).Wait(0x11a9c554)
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: 	/usr/local/go/src/sync/waitgroup.go:129 +0x84
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: vsys/mas/vendor/github.com/gammazero/nexus/client.(*Client).Close(0x11a9c500, 0x11fbbc10, 0x1)
Jul 15 12:03:09 C05AGB0000A000000297 daemon.info docker:ies-system-armhf2.2.0-87/2ec45310b45a[2257]: 	/go/src/vsys/mas/vendor/github.com/gammazero/nexus/client/client.go:760 +0x60
```

As far as I can tell, this is what's happening:

* On calls to `runHandleInvocation`, `c.activeInvHandlers.Add(1)` is called and then a goroutine is kicked off to handle the invocation.  The first thing this does is to `defer` this function:
```
		defer func() {
			c.progGateLock.Lock()
			delete(c.progGate, ctx)
			c.progGateLock.Unlock()

			c.runAction(func() {
				delete(c.invHandlerKill, msg.Request)
				c.activeInvHandlers.Done()
			})
		}()
```
* `c.runAction` `select`s between doing the action and cancelling the action if the client is closed, ie `<- c.done`.  Nothing is ever put into the channel `c.done` but it is closed, at the end of `Client.run()`.
* My guess is that there is an invocation handler running when the network connection drops.  Because of this, the invocation handler does not complete, waiting for a TCP timeout in its communication with the server.
* Then the attempt to receive at client.go:1155 fails, client.go:1157 `return`s from `run` and `c.done` is closed.  This triggers the mechanism described above where `activeInvHandlers` leaks an active goroutine.
* Once `run` exits, the code using nexus calls `Close()`.  This `Wait()`s on `activeInvHandlers` which will now never reach zero.

As for what a fix might be, I am unsure what is the best way forward.
*

**Comment by gammazero** [comment]
Thank you for your analysis and PR. I will be evaluating and testing this weekend.

To deal with this problem I am also thinking that clients should have the ability to enable websocket keep-alive (ping-pong).  Currently this is only configurable for the server.   Could add this as an option in `Config.WsCfg`
https://github.com/gammazero/nexus/blob/master/client/client.go#L78

**Comment by gammazero** [comment]
Problem summary: Calling c.runAction() here:
https://github.com/gammazero/nexus/blob/master/client/client.go#L1294
Returns immediately because the done channel is closed, without calling the cleanup function, therefore not calling `c.activeInvHandlers.Done()`.

I think it would be a sufficient fix if the call to `c.activeInvHandlers.Done()` (https://github.com/gammazero/nexus/blob/master/client/client.go#L1296) was moved outside of the function submitted to `runAction`:
```
	c.runAction(func() {
		delete(c.invHandlerKill, msg.Request)
	})
	c.activeInvHandlers.Done()
```

Can you test if that is sufficient?

**Comment by gammazero** [comment]
Submitted PR #191 to address the specific problem.

**Comment by tomkcook** [comment]
Doesn't this cause a memory leak in the case where `c.done` is closed, because then `delete(c.invHandlerKill, msg.Request)` is not called?

**Comment by gammazero** [comment]
@tomkcook I do not think it causes more of a leak than having the disconnected client around.  If `RunAction` returns without running the action, then the client is closed and cannot do anything more anyway, including leak more memory.  GC will free memory when dead client is no longer referenced.

**Comment by tomkcook** [comment]
Okay then - this looks fine to me.

**Comment by gammazero** [comment]
Fixed by #193
